File: backend/src/harness_shell_sidecar/protocol/codec.py
"""Length-prefixed protocol v1 framing."""

from pydantic import ValidationError

from .errors import FrameTooLarge, ProtocolViolation
from .models import MAX_HEADER_BYTES, MAX_PAYLOAD_BYTES, FrameEnvelope


_HEADER_DELIMITER = b"\r\n\r\n"
_CONTENT_LENGTH_PREFIX = b"Content-Length: "


def encode_frame(frame: FrameEnvelope) -> bytes:
    body = frame.model_dump_json(exclude_none=False).encode("utf-8")
    if len(body) > MAX_PAYLOAD_BYTES:
        raise FrameTooLarge(len(body), MAX_PAYLOAD_BYTES)
    return f"Content-Length: {len(body)}\r\n\r\n".encode("ascii") + body


class FrameDecoder:
    """Incrementally decode complete frames from arbitrary byte chunks."""

    def __init__(self) -> None:
        """创建一个尚未接收任何协议字节的增量解码器。"""

        # 保存尚不足以组成完整帧的跨 chunk 字节。
        self._buffer = bytearray()

    def feed(self, chunk: bytes) -> list[FrameEnvelope]:
        """追加字节并返回本次能够完整解码出的所有协议帧。"""

        self._buffer.extend(chunk)
        frames: list[FrameEnvelope] = []

        try:
            while True:
                header_end = self._buffer.find(_HEADER_DELIMITER)
                if header_end < 0:
                    if len(self._buffer) > MAX_HEADER_BYTES:
                        raise ProtocolViolation(
                            f"frame header exceeds {MAX_HEADER_BYTES} bytes"
                        )
                    return frames

                if header_end > MAX_HEADER_BYTES:
                    raise ProtocolViolation(
                        f"frame header exceeds {MAX_HEADER_BYTES} bytes"
                    )

                content_length = self._parse_header(bytes(self._buffer[:header_end]))
                if content_length > MAX_PAYLOAD_BYTES:
                    raise FrameTooLarge(content_length, MAX_PAYLOAD_BYTES)

                body_start = header_end + len(_HEADER_DELIMITER)
                frame_end = body_start + content_length
                if len(self._buffer) < frame_end:
                    return frames

                body = bytes(self._buffer[body_start:frame_end])
                del self._buffer[:frame_end]
                frames.append(self._parse_body(body))
        except (ProtocolViolation, FrameTooLarge):
            self._buffer.clear()
            raise

    @staticmethod
    def _parse_header(header: bytes) -> int:
        """严格解析单行 Content-Length 帧头并返回正文长度。"""

        lines = header.split(b"\r\n")
        if len(lines) != 1 or not lines[0].startswith(_CONTENT_LENGTH_PREFIX):
            raise ProtocolViolation(
                "frame header must contain exactly one case-sensitive Content-Length"
            )

        value = lines[0][len(_CONTENT_LENGTH_PREFIX) :]
        if not value or not value.isdigit():
            raise ProtocolViolation("Content-Length must be an unsigned decimal integer")

        return int(value)

    @staticmethod
    def _parse_body(body: bytes) -> FrameEnvelope:
        """把 UTF-8 JSON 正文校验为唯一允许的 v1 信封模型。"""

        try:
            return FrameEnvelope.model_validate_json(body)
        except (ValidationError, UnicodeDecodeError, ValueError) as exc:
            raise ProtocolViolation("frame payload is not a valid v1 envelope") from exc

File: frontend/src-tauri/src/protocol/codec.rs
use super::models::{FrameEnvelope, MAX_HEADER_BYTES, MAX_PAYLOAD_BYTES, PROTOCOL_VERSION};

const HEADER_DELIMITER: &[u8] = b"\r\n\r\n";
const CONTENT_LENGTH_PREFIX: &[u8] = b"Content-Length: ";

#[derive(Debug, Eq, PartialEq, thiserror::Error)]
pub enum ProtocolError {
    #[error("frame header exceeds {max} bytes")]
    HeaderTooLarge { max: usize },
    #[error("frame header must contain exactly one case-sensitive Content-Length")]
    InvalidHeader,
    #[error("frame payload length {actual} exceeds {max} bytes")]
    FrameTooLarge { actual: usize, max: usize },
    #[error("frame payload is not a valid v1 envelope")]
    InvalidEnvelope,
}

pub fn encode_frame(frame: &FrameEnvelope) -> Result<Vec<u8>, ProtocolError> {
    if frame.protocol_version != PROTOCOL_VERSION || frame.sequence == 0 {
        return Err(ProtocolError::InvalidEnvelope);
    }
    let body = serde_json::to_vec(frame).map_err(|_| ProtocolError::InvalidEnvelope)?;
    if body.len() > MAX_PAYLOAD_BYTES {
        return Err(ProtocolError::FrameTooLarge {
            actual: body.len(),
            max: MAX_PAYLOAD_BYTES,
        });
    }

    let mut encoded = format!("Content-Length: {}\r\n\r\n", body.len()).into_bytes();
    encoded.extend(body);
    Ok(encoded)
}

#[derive(Default)]
pub struct FrameDecoder {
    buffer: Vec<u8>,
}

impl FrameDecoder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn push(&mut self, chunk: &[u8]) -> Result<Vec<FrameEnvelope>, ProtocolError> {
        self.buffer.extend_from_slice(chunk);
        let result = self.decode_available();
        if result.is_err() {
            self.buffer.clear();
        }
        result
    }

    fn decode_available(&mut self) -> Result<Vec<FrameEnvelope>, ProtocolError> {
        let mut frames = Vec::new();
        loop {
            let Some(header_end) = find_subslice(&self.buffer, HEADER_DELIMITER) else {
                if self.buffer.len() > MAX_HEADER_BYTES {
                    return Err(ProtocolError::HeaderTooLarge {
                        max: MAX_HEADER_BYTES,
                    });
                }
                return Ok(frames);
            };

            if header_end > MAX_HEADER_BYTES {
                return Err(ProtocolError::HeaderTooLarge {
                    max: MAX_HEADER_BYTES,
                });
            }
            let content_length = parse_header(&self.buffer[..header_end])?;
            if content_length > MAX_PAYLOAD_BYTES {
                return Err(ProtocolError::FrameTooLarge {
                    actual: content_length,
                    max: MAX_PAYLOAD_BYTES,
                });
            }

            let body_start = header_end + HEADER_DELIMITER.len();
            let frame_end = body_start + content_length;
            if self.buffer.len() < frame_end {
                return Ok(frames);
            }

            let body = self.buffer[body_start..frame_end].to_vec();
            self.buffer.drain(..frame_end);
            let frame =
                serde_json::from_slice(&body).map_err(|_| ProtocolError::InvalidEnvelope)?;
            frames.push(frame);
        }
    }
}

fn parse_header(header: &[u8]) -> Result<usize, ProtocolError> {
    if header.contains(&b'\r')
        || header.contains(&b'\n')
        || !header.starts_with(CONTENT_LENGTH_PREFIX)
    {
        return Err(ProtocolError::InvalidHeader);
    }

    let value = &header[CONTENT_LENGTH_PREFIX.len()..];
    if value.is_empty() || !value.iter().all(u8::is_ascii_digit) {
        return Err(ProtocolError::InvalidHeader);
    }
    std::str::from_utf8(value)
        .ok()
        .and_then(|value| value.parse().ok())
        .ok_or(ProtocolError::InvalidHeader)
}

fn find_subslice(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    haystack
        .windows(needle.len())
        .position(|window| window == needle)
}
